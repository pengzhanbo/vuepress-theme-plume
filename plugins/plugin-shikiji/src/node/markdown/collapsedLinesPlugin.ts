import type { Markdown } from 'vuepress/markdown'
import { resolveCollapsedLines } from '../utils/index.js'

export interface MarkdownItCollapsedLinesOptions {
  /**
   * Whether to collapse code blocks when they exceed a certain number of lines,
   *
   * - If `number`, collapse starts from line `number`.
   * - If `true`, collapse starts from line 15 by default.
   * - If `false`, disable collapse.
   * @default false
   */
  collapsedLines?: boolean | number
}

export function collapsedLinesPlugin(md: Markdown, {
  collapsedLines: collapsedLinesOptions = false,
}: MarkdownItCollapsedLinesOptions = {}): void {
  const rawFence = md.renderer.rules.fence!

  md.renderer.rules.fence = (...args) => {
    const [tokens, index] = args
    const token = tokens[index]
    // get token info
    const info = token.info ? md.utils.unescapeAll(token.info).trim() : ''
    const code = rawFence(...args)

    // resolve collapsed-lines mark from token info
    const collapsedLinesInfo
      = resolveCollapsedLines(info) ?? collapsedLinesOptions

    if (collapsedLinesInfo === false) {
      return code
    }

    const lines
      = code.slice(code.indexOf('<code>'), code.indexOf('</code>')).split('\n').length

    const startLines
      = typeof collapsedLinesInfo === 'number' ? collapsedLinesInfo : 15

    if (lines < startLines) {
      return code
    }

    const collapsedLinesCode = `<div class="collapsed-lines"></div>`
    const styles = `--vp-collapsed-lines:${startLines};`

    const finalCode = code
      .replace(/<\/div>$/, `${collapsedLinesCode}</div>`)
      .replace(/"(language-[^"]*)"/, '"$1 has-collapsed collapsed"')
      .replace(/^<div[^>]*>/, (match) => {
        if (!match.includes('style=')) {
          return `${match.slice(0, -1)} style="${styles}">`
        }
        return match.replace(/(style=")/, `$1${styles}`)
      })

    return finalCode
  }
}
