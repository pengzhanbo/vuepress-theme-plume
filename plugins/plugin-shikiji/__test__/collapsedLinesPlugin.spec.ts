import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { collapsedLinesPlugin } from '../src/node/markdown/collapsedLinesPlugin.js'

const FENCE = '```'

function createMarkdown(collapsedLines?: boolean | number, hasStyles = false) {
  const md = new MarkdownIt()

  md.options.highlight = str => `<pre><code>${str}</code></pre>`
  const fence = md.renderer.rules.fence!
  md.renderer.rules.fence = (...args) =>
    `<div class="language-"${hasStyles ? ' style="color: red;"' : ''}>${fence(...args)}</div>`

  md.use(collapsedLinesPlugin, collapsedLines ? { collapsedLines } : undefined)

  return md
}

describe('collapsedLinesPlugin', () => {
  const genLines = (n: number) =>
    Array.from({ length: n }).map((_, i) => `const line${i + 1} = ${i + 1}`).join('\n')
  const code = `\
${FENCE}
${genLines(10)}
${FENCE}

${FENCE}js
${genLines(20)}
${FENCE}

${FENCE}js :collapsed-lines
${genLines(20)}
${FENCE}

${FENCE}js :collapsed-lines=10
${genLines(12)}
${FENCE}

${FENCE}js :no-collapsed-lines
${genLines(20)}
${FENCE}
`
  it('should work with default', () => {
    const md = createMarkdown()
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should work with true', () => {
    const md = createMarkdown(true)
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should work with number', () => {
    const md = createMarkdown(10)
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work with includes styles', () => {
    const md = createMarkdown(true, true)
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should not work with false', () => {
    const md = createMarkdown(false)
    expect(md.render(code)).toMatchSnapshot()
  })
})
