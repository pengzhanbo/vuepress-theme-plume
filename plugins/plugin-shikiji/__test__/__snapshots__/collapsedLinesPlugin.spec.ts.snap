// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`collapsedLinesPlugin > should not work with false 1`] = `
"<div class="language-"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
</code></pre>
</div><div class="language-"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
</div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:15;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
<div class="collapsed-lines"></div></div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:10;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
</code></pre>
<div class="collapsed-lines"></div></div><div class="language-"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
</div>"
`;

exports[`collapsedLinesPlugin > should not work with includes styles 1`] = `
"<div class="language-" style="color: red;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
</code></pre>
</div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:15;color: red;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
<div class="collapsed-lines"></div></div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:15;color: red;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
<div class="collapsed-lines"></div></div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:10;color: red;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
</code></pre>
<div class="collapsed-lines"></div></div><div class="language-" style="color: red;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
</div>"
`;

exports[`collapsedLinesPlugin > should work with default 1`] = `
"<div class="language-"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
</code></pre>
</div><div class="language-"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
</div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:15;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
<div class="collapsed-lines"></div></div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:10;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
</code></pre>
<div class="collapsed-lines"></div></div><div class="language-"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
</div>"
`;

exports[`collapsedLinesPlugin > should work with number 1`] = `
"<div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:10;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
</code></pre>
<div class="collapsed-lines"></div></div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:10;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
<div class="collapsed-lines"></div></div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:15;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
<div class="collapsed-lines"></div></div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:10;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
</code></pre>
<div class="collapsed-lines"></div></div><div class="language-"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
</div>"
`;

exports[`collapsedLinesPlugin > should work with true 1`] = `
"<div class="language-"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
</code></pre>
</div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:15;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
<div class="collapsed-lines"></div></div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:15;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
<div class="collapsed-lines"></div></div><div class="language- has-collapsed collapsed" style="--vp-collapsed-lines:10;"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
</code></pre>
<div class="collapsed-lines"></div></div><div class="language-"><pre><code>const line1 = 1
const line2 = 2
const line3 = 3
const line4 = 4
const line5 = 5
const line6 = 6
const line7 = 7
const line8 = 8
const line9 = 9
const line10 = 10
const line11 = 11
const line12 = 12
const line13 = 13
const line14 = 14
const line15 = 15
const line16 = 16
const line17 = 17
const line18 = 18
const line19 = 19
const line20 = 20
</code></pre>
</div>"
`;
